### PR TITLE
Fixing and improve the SPN recording demo

### LINKED ISSUE #308: Fixing and improve User experience of SPN Recorder demo
The max record length is 10s, but the recording can go infinitely. It should be auto paused. 
Since we agree to expend the max recording length to 5 mins, so we need to update this demo too. 
Edit the description of this demo to make the user understand what's is this demo stands for.

### PR DESCRIPTION
**Pull request by TenghuiZhang**
Fix #308 
Now the SPN recording demo is behave the same as Worklet recording demo
This PR also fix the format issue of github config file(main.yml).

### CODE REVIEW

File: src/audio-worklet/migration/spn-recorder/app.js [hunk 1]
```diff
@@ -11,10 +11,17 @@ const context = new AudioContext();
 // Arbitrary buffer size, not specific for a reason
 const BUFFER_SIZE = 256;
 let recordingLength = 0;
+let recordBuffer = [[], []];
 let isRecording = false;
 let isMonitoring = false;
 let visualizationEnabled = true;
 
+let recordButton = document.querySelector('#record');
+let recordText = recordButton.querySelector('span');
```
Comment (hoch) [inline src/audio-worklet/migration/spn-recorder/app.js:20]: I recommend using `id="record-text"` instead of selecting `span`. There could be others.

File: src/audio-worklet/migration/spn-recorder/app.js [hunk 2]
```diff
@@ -90,8 +101,8 @@ function setupScriptProcessor(recordBuffer, passSampleToVisualizers) {
 
       // While recording, feed data to recording buffer at the proper time.
       if (isRecording) {
-        recordBuffer
-            .copyToChannel(currentSamples[channel], channel, recordingLength);
+        let frameNumber = recordingLength / BUFFER_SIZE;
+        recordBuffer[channel][frameNumber] = currentSamples[channel].slice();
```
Comment (hoch) [inline src/audio-worklet/migration/spn-recorder/app.js:105]: What does this do?

Also `.slice()` creates a new instance. Is this intended?

File: src/audio-worklet/migration/spn-recorder/app.js [hunk 3]
```diff
@@ -101,6 +112,21 @@ function setupScriptProcessor(recordBuffer, passSampleToVisualizers) {
     // Update tracked recording length.
     if (isRecording) {
       recordingLength += BUFFER_SIZE;
+      
+      if (recordingLength > recordingProperties.maxFrameCount) {
+        isRecording = !isRecording;
+
+        const finalRecordBuffer =
+            createFinalRecordBuffer(recordingProperties);
+        const wavUrl = createLinkFromAudioBuffer(finalRecordBuffer, true);
```
Comment (hoch) [inline src/audio-worklet/migration/spn-recorder/app.js:121]: Let's rename `wavUrl` to `audioFileUrl`.

File: src/audio-worklet/migration/spn-recorder/app.js [hunk 4]
```diff
@@ -101,6 +112,21 @@ function setupScriptProcessor(recordBuffer, passSampleToVisualizers) {
...
+        player.src = wavUrl;
+        downloadLink.href = wavUrl;
+        downloadLink.download =
+            `recording-${new Date().getMilliseconds().toString()}.wav`;
+        downloadButton.disabled = false;
+        recordText.innerHTML = 'Ready to download 5 mins';
```
Comment (hoch) [inline src/audio-worklet/migration/spn-recorder/app.js:127]: `innerHTML` to `textContent`.

Also are we downloading a "5-minute file" always?

File: src/audio-worklet/migration/spn-recorder/app.js [hunk 5]
```diff
@@ -305,3 +325,25 @@ function setupRecordingGainVis() {
...
 }
+
+/**
+ * Create the recording buffer with right size
+ * @param {object} recordingProperties Properties of record buffer
+ * @return {AudioBuffer} Record buffer for current recording
+ */
+const createFinalRecordBuffer = (recordingProperties) => {
+  const contextRecordBuffer = context.createBuffer(
+      2, recordingLength, context.sampleRate,);
```
Comment (hoch) [inline src/audio-worklet/migration/spn-recorder/app.js:336]: One redundant comma at the end.

File: src/audio-worklet/migration/spn-recorder/app.js [hunk 6]
```diff
@@ -305,3 +325,25 @@ function setupRecordingGainVis() {
...
+  let startPos = 0;
```
Comment (hoch) [inline src/audio-worklet/migration/spn-recorder/app.js:337]: Avoid abbreviation if possible: `startPosition`.

Also "position" is rather abstract. Is this an index of an array? Then say `startIndex` and add comments for it.

File: src/audio-worklet/migration/spn-recorder/app.js [hunk 7]
```diff
@@ -305,3 +325,25 @@ function setupRecordingGainVis() {
...
+
+  for (let frame = 0; frame < recordBuffer[0].length; frame++){
+    for (let channel = 0; 
+        channel < recordingProperties.numberOfChannels; channel++) {
+
```
Comment (hoch) [inline src/audio-worklet/migration/spn-recorder/app.js:342]: A blank line here doesn't seem necessary.

File: src/audio-worklet/migration/spn-recorder/app.js [hunk 8]
```diff
@@ -305,3 +325,25 @@ function setupRecordingGainVis() {
...
+ * @return {AudioBuffer} Record buffer for current recording
+ */
+const createFinalRecordBuffer = (recordingProperties) => {
+  const contextRecordBuffer = context.createBuffer(
+      2, recordingLength, context.sampleRate,);
+  let startPos = 0;
+
+  for (let frame = 0; frame < recordBuffer[0].length; frame++){
+    for (let channel = 0; 
+        channel < recordingProperties.numberOfChannels; channel++) {
```
Comment (hoch) [inline src/audio-worklet/migration/spn-recorder/app.js:341]: I'd do

```js
for (let channel = 0;
     channel < recordingProperties.numberOfChannels;
     channel++) {
```

**Review by hoch — COMMENTED** [review]
Getting there! I made some comments.

File: src/audio-worklet/migration/spn-recorder/app.js (hunk 2 above)
Comment (TenghuiZhang) [reply src/audio-worklet/migration/spn-recorder/app.js:105]: "=" in js make Object, array a shallow copy. We have to use slice to create a deep copy. Therefore, the data in recordingBuffer will be modified based on currentSamples[channel] update. 

There is another way is copy the float number in array one by one since "=" in js make number a deep copy. Due to the time and readability consuming, I think using 'slice' is better.

File: src/audio-worklet/migration/spn-recorder/app.js (hunk 4 above)
Comment (TenghuiZhang) [reply src/audio-worklet/migration/spn-recorder/app.js:127]: No, the newest algorithm can make the download always be the right length.

File: src/audio-worklet/migration/spn-recorder/app.js (hunk 2 above)
Comment (hoch) [reply src/audio-worklet/migration/spn-recorder/app.js:105]: Understood, but since this is happening around Float32Array, I recommend using `.set()` method for faster operation: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray/set

It might need a bit or writing, but could you give some thoughts?

File: src/audio-worklet/migration/spn-recorder/app.js (hunk 4 above)
Comment (hoch) [reply src/audio-worklet/migration/spn-recorder/app.js:127]: Then we need to change the text to reflect what's happening.

File: src/audio-worklet/migration/spn-recorder/app.js (hunk 2 above)
Comment (TenghuiZhang) [reply src/audio-worklet/migration/spn-recorder/app.js:105]: Based on the reference, we only need to assign a new Float32Array. 

recordBuffer[channel][frameNumber] = new Float32Array(currentSamples[channel]);

In this way, we don't need set() OR slice(). I think these two way speed are similar. Basically, we are creating a new float32array.

File: .github/workflows/main.yml [hunk 9]
```diff
@@ -15,7 +15,7 @@ jobs:
     runs-on: ubuntu-latest
     steps:
     - name: Check out this repo
-      uses: actions/checkout@v2
+      uses: actions/checkout@a81bbbf8298c0fa03ea29cdc473d45769f953675
```
Comment (hoch) [inline .github/workflows/main.yml:18]: This doesn't seem relevant to this change. Can you explain?

File: src/audio-worklet/migration/worklet-recorder/app.js [hunk 10]
```diff
@@ -31,17 +31,15 @@ async function init() {
...
   });
 
   const micSourceNode = context.createMediaStreamSource(micStream);
 
   const recordingProperties = {
     numberOfChannels: micSourceNode.channelCount,
     sampleRate: context.sampleRate,
-    maxFrameCount: context.sampleRate * 300,
-  };
+    maxFrameCount: context.sampleRate * 300};
```
Comment (hoch) [inline src/audio-worklet/migration/worklet-recorder/app.js:42]: Let's move the ending curly brace to the next line.

**Review by hoch — APPROVED** [review]
LGTM with nits

File: .github/workflows/main.yml (hunk 9 above)
Comment (TenghuiZhang) [reply .github/workflows/main.yml]: I am still in the process of figuring it out. The reason I am doing it is to pass the github check. We have this issue before, but at that time, I pass the test by changing the checkout@v3 to checkout@v2 and add the permission:write. This time, the presubmit is blocked again. Haven't find a right solution yet.
